docs/engineering-principles.md

```diff
@@ -0,0 +1,48 @@
+
+### The Principle of Core Purity
+
+Whenever possible prefer small, pure functions. Pure functions are those that given the same inputs will always return the same output, and that have no side effects. Where side effects must occur, keep the core of the application code pure and move imperative methods to the outer edges of the application layer.
+
+
+### The Principle of Incremental Change
+
+To minimize the risk to the stability and security of the application, changes to the code should be incremental and as small as possible. Large changes to the codebase become difficult to review effectively and make the commit history less itemized. Furthermore, incremental change promotes better test coverage and makes it easier to require tests to cover the introduced changes.
+
+### The Principle of Incremental Decentralization
+
+We may sometimes ship with centralized services by default, but if we must, users should be able to opt out of any of these during onboarding, ideally easily swapping them out for decentralized alternatives. When we build features that rely upon centralized services we must build them in such a way where the experience isn't completely broken when the service is disrupted. We are trying to build a decentralized future and we must make choices that move us directionally towards those goals.
+
+### The Principle of Self-Soverign Identity
+
+We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right we must treat those features as opt in, not opt out. When we do build features that make user data known to us we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity protecting way. This principle must be extended to cover all third party code as well.
+
+### The Principle of Auditability
+
+All code contributions must be reviewed thoroughly for code quality, usability and adherence to these principles. Code contributions must be kept as small as possible (see Principle of Incremental Change) to promote auditability. Avoid long running feature branches because they are not auditable, and pose a risk to the stability and security of the product.
+
+
+### The Principle of "Be Methodical and Fix Things"
+
+We should prioritize fixing bugs and improving stability over creating new features whenever possible. All code generates new bugs and builds technical debt, this is inevitable. However, if we focus on building our features with stability in mind and throw out the "Move Fast and Break Things" mentality of most web 2.0 applications we can minimize the amount of bugs we introduce. Furthermore, whenever bugs are known to exist we should expend the necessary energy to remedy them. 
+
+### The Principle of Ecosystem Momentum
```

Comment:
I don't understand what "ecosystem momentum" means. In the description momentum is referenced in one line, which I also don't understand:

>We must view these breaking changes as a potential reset of our momentum within the ecosystem.

Otherwise the text here seems to primarily talk about API stability, and the challenges and consequences of breaking changes.

Maybe it refers to the ecosystem's rate of growth? The rate of growth in our userbase, or our developer/dapp userbase? The growth of the "MetaMask-compatible dapps" ecosystem?

Reply:
I was taking it to mean the MetaMask compatible dapps ecosystem, which I could either spell out to further clarify, or we could consider an entire reword of this principle. Do you feel strongly one way or the other?

Reply:
One way I am looking at "ecosystem momentum" is given that momentum = mass x velocity, I am thinking ecosystem momentum = dapp compatibility x velocity. The velocity would be correlated with the API stability and our ability to mitigate downtime with any breaking changes. Maybe another way to say
> We must view these breaking changes as a potential reset of our momentum within the ecosystem.

could be

> We must view these breaking changes as a hindrance to our momentum within the ecosystem.

I am liking this "The Principle of Ecosystem Momentum" proposal to keep us accountable as providers in the ecosystem. Maybe we can make it clearer that this is in regards to the dapps ecosystem. Links might also be helpful here.

Reply:
I think it would be clearer if the title more directly talked about not breaking dapps. Something like Linux's "Don't break userspace", or "Don't break the web" for browser changes. Those are nice and succinct, easy to remember, and I often hear them repeated in discussions.

Reply:
+1 to make the title clearer (after my overthinking above 😄). Some other suggestions: "The Principle of Reliable Dapp Support" or "The Principle of Providing Dapp Stability"

Reply:
I went with "principle of reliability" and added feature deprecation as a thing we should treat carefully as well. Open to tinkering the name with other options

Reply:
@Gudahtt @digiwand is this comment line resolved or do we need further revisions?

Reply:
I think it's OK for now